# Predicting footballer player wage based on respective FIFA / EAFC plater stats

The problem we are addressing is correctly evaluating a playe's wage based on their stats. With ever more growing talent in the field, it is vital as a football club to accurately price your players to stay competitive. Our goal is to study & build machine learning models to accurately predict a footballer's wage based on a dataset of players wage and stats.

# Importing data

We are going to be utilising the [EA Sports FC 24 complete player dataset](https://www.kaggle.com/datasets/omkargowda/football-players-stats-premier-league-20212022/) from kaggle for player stats and valautions.

In [48]:
import pandas as pd
# instructing pandas to not truncate column widths when displaying interactively
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', None)

datasets_fp = 'datasets/'
encoding = 'unicode_escape'
df = pd.read_csv(datasets_fp+'male_players.csv', encoding=encoding)
df

C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\646558934.py:8: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(datasets_fp+'male_players.csv', encoding=encoding)


player_id                                             player_url  \
0          231747                    /player/231747/kylian-mbappe/240002   
1          239085                   /player/239085/erling-haaland/240002   
2          192985                  /player/192985/kevin-de-bruyne/240002   
3          158023                     /player/158023/lionel-messi/240002   
4          165153                    /player/165153/karim-benzema/240002   
...           ...                                                    ...   
180016     220806   https://sofifa.com/player/220806/ellis-redman/150002   
180017     225509  https://sofifa.com/player/225509/aaron-collins/150002   
180018     217591    https://sofifa.com/player/217591/piotr-zemlo/150002   
180019     222947     https://sofifa.com/player/222947/tom-davies/150002   
180020     201197   https://sofifa.com/player/201197/paul-tisdale/150002   

        fifa_version  fifa_update update_as_of    short_name  \
0               24.0          2.0   2023-09-22    K. Mbappé   
1               24.0          2.0   2023-09-22    E. Haaland   
2               24.0          2.0   2023-09-22  K. De Bruyne   
3               24.0          2.0   2023-09-22      L. Messi   
4               24.0          2.0   2023-09-22    K. Benzema   
...              ...          ...          ...           ...   
180016          15.0          2.0   2014-09-18     E. Redman   
180017          15.0          2.0   2014-09-18    A. Collins   
180018          15.0          2.0   2014-09-18    P. Żemło   
180019          15.0          2.0   2014-09-18     T. Davies   
180020          15.0          2.0   2014-09-18    P. Tisdale   

                              long_name player_positions  overall  potential  \
0                 Kylian Mbappé Lottin           ST, LW       91         94   
1                  Erling Braut Haaland               ST       91         94   
2                       Kevin De Bruyne          CM, CAM       91         91   
3       Lionel Andrés Messi Cuccittini          CF, CAM       90         90   
4                         Karim Benzema           CF, ST       90         90   
...                                 ...              ...      ...        ...   
180016                     Ellis Redman               CB       41         61   
180017        Aaron Graham John Collins               ST       41         50   
180018                    Piotr Żemło           LM, LB       40         50   
180019        Thomas Christopher Davies               CB       40         49   
180020                     Paul Tisdale          CM, CAM       40         40   

          value_eur  wage_eur  age         dob  height_cm  weight_kg  \
0       181500000.0  230000.0   24  1998-12-20        182         75   
1       185000000.0  340000.0   22  2000-07-21        195         94   
2       103000000.0  350000.0   32  1991-06-28        181         75   
3        41000000.0   23000.0   36  1987-06-24        169         67   
4        51000000.0   95000.0   35  1987-12-19        185         81   
...             ...       ...  ...         ...        ...        ...   
180016      20000.0    2000.0   17  1996-09-26        187         81   
180017      30000.0    2000.0   17  1997-06-01        178         57   
180018      15000.0    2000.0   18  1995-07-10        190         76   
180019      15000.0    2000.0   22  1992-04-18        180         70   
180020          NaN    2000.0   41  1973-01-14        175         72   

        club_team_id            club_name  league_id          league_name  \
0               73.0  Paris Saint Germain       16.0              Ligue 1   
1               10.0      Manchester City       13.0       Premier League   
2               10.0      Manchester City       13.0       Premier League   
3           112893.0          Inter Miami       39.0  Major League Soccer   
4              607.0           Al Ittihad      350.0           Pro League   
...              ...                  

Let's have a quick look at the dataframe

In [49]:
df.head(5) # first 5 rows

player_id                             player_url  fifa_version  \
0     231747    /player/231747/kylian-mbappe/240002          24.0   
1     239085   /player/239085/erling-haaland/240002          24.0   
2     192985  /player/192985/kevin-de-bruyne/240002          24.0   
3     158023     /player/158023/lionel-messi/240002          24.0   
4     165153    /player/165153/karim-benzema/240002          24.0   

   fifa_update update_as_of    short_name                        long_name  \
0          2.0   2023-09-22    K. Mbappé            Kylian Mbappé Lottin   
1          2.0   2023-09-22    E. Haaland             Erling Braut Haaland   
2          2.0   2023-09-22  K. De Bruyne                  Kevin De Bruyne   
3          2.0   2023-09-22      L. Messi  Lionel Andrés Messi Cuccittini   
4          2.0   2023-09-22    K. Benzema                    Karim Benzema   

  player_positions  overall  potential    value_eur  wage_eur  age  \
0           ST, LW       91         94  181500000.0  230000.0   24   
1               ST       91         94  185000000.0  340000.0   22   
2          CM, CAM       91         91  103000000.0  350000.0   32   
3          CF, CAM       90         90   41000000.0   23000.0   36   
4           CF, ST       90         90   51000000.0   95000.0   35   

          dob  height_cm  weight_kg  club_team_id            club_name  \
0  1998-12-20        182         75          73.0  Paris Saint Germain   
1  2000-07-21        195         94          10.0      Manchester City   
2  1991-06-28        181         75          10.0      Manchester City   
3  1987-06-24        169         67      112893.0          Inter Miami   
4  1987-12-19        185         81         607.0           Al Ittihad   

   league_id          league_name  league_level club_position  \
0       16.0              Ligue 1           1.0            LW   
1       13.0       Premier League           1.0            ST   
2       13.0       Premier League           1.0           SUB   
3       39.0  Major League Soccer           1.0            RF   
4      350.0           Pro League           1.0            RS   

   club_jersey_number club_loaned_from club_joined_date  \
0                 7.0              NaN       2018-07-01   
1                 9.0              NaN       2022-07-01   
2                17.0              NaN       2015-08-30   
3                10.0              NaN       2023-07-16   
4                 9.0              NaN       2023-07-01   

   club_contract_valid_until_year  nationality_id nationality_name  \
0                          2024.0              18           France   
1                          2027.0              36           Norway   
2                          2025.0               7          Belgium   
3                          2025.0              52        Argentina   
4                          2026.0              18           France   

   nation_team_id nation_position  nation_jersey_number preferred_foot  \
0          1335.0              LW                  10.0          Right   
1          1352.0              ST                   9.0           Left   
2          1325.0             CAM                   7.0          Right   
3          1369.0              RW                  10.0           Left   
4             NaN             NaN                   NaN          Right   

   weak_foot  skill_moves  international_reputation      work_rate  \
0          4            5                         5       High/Low   
1          3            3                         5    High/Medium   
2          5            4                         5    High/Medium   
3          4            4                         5        Low/Low   
4          4            4                         5  Medium/Medium   

          body_type real_face  release_clause_eur  \
0            Unique       Yes         349400000.0   
1            Unique       Yes         356100000.0   
2            Unique       Yes         190600000.0   
3       

We will focusing on the Premier League data only so let's filter the dataset to satisfy this requirement

In [50]:
df = df[df['league_name'] == 'Premier League'] # filter league_name to 'Premier League'
df

player_id  \
1          239085   
2          192985   
9          212831   
10         231866   
11         239818   
...           ...   
179683     225068   
179693     222112   
179720     201214   
179783     213693   
179819     222111   

                                                         player_url  \
1                              /player/239085/erling-haaland/240002   
2                             /player/192985/kevin-de-bruyne/240002   
9                       /player/212831/alisson-ramses-becker/240002   
10                 /player/231866/rodrigo-hernandez-cascante/240002   
11               /player/239818/ruben-santos-gato-alves-dias/240002   
...                                                             ...   
179683  https://sofifa.com/player/225068/jamie-dean-mcdonagh/150002   
179693       https://sofifa.com/player/222112/elliot-justham/150002   
179720        https://sofifa.com/player/201214/apti-akhyadov/150002   
179783         https://sofifa.com/player/213693/joe-rothwell/150002   
179819           https://sofifa.com/player/222111/andy-parry/150002   

        fifa_version  fifa_update update_as_of    short_name  \
1               24.0          2.0   2023-09-22    E. Haaland   
2               24.0          2.0   2023-09-22  K. De Bruyne   
9               24.0          2.0   2023-09-22       Alisson   
10              24.0          2.0   2023-09-22         Rodri   
11              24.0          2.0   2023-09-22   Rúben Dias   
...              ...          ...          ...           ...   
179683          15.0          2.0   2014-09-18   J. McDonagh   
179693          15.0          2.0   2014-09-18    E. Justham   
179720          15.0          2.0   2014-09-18   A. Akhyadov   
179783          15.0          2.0   2014-09-18   J. Rothwell   
179819          15.0          2.0   2014-09-18      A. Parry   

                                long_name player_positions  overall  \
1                    Erling Braut Haaland               ST       91   
2                         Kevin De Bruyne          CM, CAM       91   
9                  Alisson Ramsés Becker               GK       89   
10            Rodrigo Hernández Cascante          CDM, CM       89   
11      Rúben dos Santos Gato Alves Dias               CB       89   
...                                   ...              ...      ...   
179683                     Jamie McDonagh               RW       50   
179693                     Elliot Justham               GK       50   
179720                      Apti Akhyadov               CF       49   
179783                       Joe Rothwell               CM       49   
179819                       Andrew Parry      CDM, CB, RB       49   

        potential    value_eur  wage_eur  age         dob  height_cm  \
1              94  185000000.0  340000.0   22  2000-07-21        195   
2              91  103000000.0  350000.0   32  1991-06-28        181   
9              90   77500000.0  180000.0   30  1992-10-02        193   
10             90  105500000.0  250000.0   27  1996-06-22        191   
11             90  106500000.0  250000.0   26  1997-05-14        187   
...           ...          ...       ...  ...         ...        ...   
179683         55      30000.0    2000.0   18  1996-05-08        168   
179693         54      30000.0    2000.0   24  1990-02-18        191   
179720         61      30000.0    2000.0   20  1993-08-24        173   
179783         55      25000.0    2000.0   19  1995-01-11        185   
179819         53      20000.0    2000.0   22  1991-09-13        179   

        weight_kg  club_team_id          club_name  league_id     league_name  \
1              94          10.0    Manchester City       13.0  Premier League   
2              75          10.0    Manchester City       13.0  Premier League   
9              91           9.0          Liverpool       13.0  Premier League   
10             82          10.0    Manchester City       13.0  Premier League   
11        

# EDA

Now we have sucessfully imported the dataset, we need to investigate and process the data to make it suitable to a machine learning model.

Let's do a quick analysis of the data and it's columns using the pandas describe() method

In [51]:
df.describe(include='all') # using include='all' to show us all columns

player_id                            player_url  fifa_version  \
count     9371.000000                                  9371   9371.000000   
unique            NaN                                  9371           NaN   
top               NaN  /player/239085/erling-haaland/240002           NaN   
freq              NaN                                     1           NaN   
mean    206326.221001                                   NaN     18.991356   
std      37041.003155                                   NaN      2.827565   
min         27.000000                                   NaN     15.000000   
25%     188566.500000                                   NaN     17.000000   
50%     208797.000000                                   NaN     19.000000   
75%     231030.000000                                   NaN     21.000000   
max     278102.000000                                   NaN     24.000000   

        fifa_update update_as_of    short_name long_name player_positions  \
count        9371.0         9371          9371      9371             9371   
unique          NaN           10          3536      3607              570   
top             NaN   2017-09-18  A. Miranchuk   Ben Mee               CB   
freq            NaN         1169            13        10             1196   
mean            2.0          NaN           NaN       NaN              NaN   
std             0.0          NaN           NaN       NaN              NaN   
min             2.0          NaN           NaN       NaN              NaN   
25%             2.0          NaN           NaN       NaN              NaN   
50%             2.0          NaN           NaN       NaN              NaN   
75%             2.0          NaN           NaN       NaN              NaN   
max             2.0          NaN           NaN       NaN              NaN   

            overall    potential     value_eur       wage_eur          age  \
count   9371.000000  9371.000000  9.359000e+03    9371.000000  9371.000000   
unique          NaN          NaN           NaN            NaN          NaN   
top             NaN          NaN           NaN            NaN          NaN   
freq            NaN          NaN           NaN            NaN          NaN   
mean      70.512752    75.696190  7.316704e+06   36509.065201    25.073738   
std        8.085491     6.556593  1.367520e+07   45322.942691     4.520321   
min       44.000000    53.000000  2.000000e+04     500.000000    16.000000   
25%       65.000000    71.000000  6.250000e+05    5000.000000    21.000000   
50%       71.000000    76.000000  2.100000e+06   20000.000000    25.000000   
75%       76.000000    80.000000  8.000000e+06   50000.000000    28.000000   
max       91.000000    94.000000  1.850000e+08  370000.000000    43.000000   

               dob    height_cm    weight_kg   club_team_id  \
count         9371  9371.000000  9371.000000    9371.000000   
unique        2914          NaN          NaN            NaN   
top     1986-02-05          NaN          NaN            NaN   
freq            24          NaN          NaN            NaN   
mean           NaN   182.041831    75.623306   32217.889339   
std            NaN     7.030029     7.427244   48547.678082   
min            NaN   160.000000    51.000000       1.000000   
25%            NaN   177.000000    70.000000      13.000000   
50%            NaN   182.000000    75.000000    1794.000000   
75%            NaN   187.000000    80.000000  100767.000000   
max            NaN   203.000000   101.000000  113704.000000   

                club_name    league_id     league_name  league_level  \
count                9371  9371.000000            9371   9179.000000   
unique                 47          NaN               1           NaN   
top     Manchester United          NaN  Premier League           NaN   
freq                  327          NaN            9371           NaN   
mean                  NaN    57.213211             NaN      1.206232   
std                   NaN    92.9235

Let's find out exactly how many columns we have

In [52]:
len(df.columns)

109

At an initial glance I've observed not all 109 columns in the dataset are relevant, so let's drop the columns we won't need. I decided on which columns to drop based on if the data is likely to affect a player's performance / value.

In [53]:
df.drop([
    'player_url',
    'update_as_of',
    'short_name',
    'long_name',
    'dob',
    'club_name',
    'league_name',
    'club_position',
    'club_loaned_from',
    'club_joined_date',
    'nationality_id',
    'nation_position',
    'body_type',
    'real_face',
    'player_tags',
    'player_traits',
    'fifa_version',
    'fifa_update',
    'league_level',
    'club_team_id',
    'league_id',
    'club_jersey_number',
    'club_contract_valid_until_year',
    'nation_team_id',
    'nation_jersey_number',
    'release_clause_eur',
    'ls',
    'st',
    'rs',
    'lw',
    'lf',
    'cf',
    'rf',
    'rw',
    'lam',
    'cam',
    'ram',
    'lm',
    'lcm',
    'cm',
    'rcm',
    'rm',
    'lwb',
    'ldm',
    'cdm',
    'rdm',
    'rwb',
    'lb',
    'lcb',
    'cb',
    'rcb',
    'rb',
    'gk',
], axis=1, inplace=True)
df

C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\3888672542.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop([


player_id player_positions  overall  potential    value_eur  wage_eur  \
1          239085               ST       91         94  185000000.0  340000.0   
2          192985          CM, CAM       91         91  103000000.0  350000.0   
9          212831               GK       89         90   77500000.0  180000.0   
10         231866          CDM, CM       89         90  105500000.0  250000.0   
11         239818               CB       89         90  106500000.0  250000.0   
...           ...              ...      ...        ...          ...       ...   
179683     225068               RW       50         55      30000.0    2000.0   
179693     222112               GK       50         54      30000.0    2000.0   
179720     201214               CF       49         61      30000.0    2000.0   
179783     213693               CM       49         55      25000.0    2000.0   
179819     222111      CDM, CB, RB       49         53      20000.0    2000.0   

        age  height_cm  weight_kg     nationality_name preferred_foot  \
1        22        195         94               Norway           Left   
2        32        181         75              Belgium          Right   
9        30        193         91               Brazil          Right   
10       27        191         82                Spain          Right   
11       26        187         82             Portugal          Right   
...     ...        ...        ...                  ...            ...   
179683   18        168         58  Republic of Ireland          Right   
179693   24        191         80              England           Left   
179720   20        173         68               Russia          Right   
179783   19        185         77              England          Right   
179819   22        179         73              England          Right   

        weak_foot  skill_moves  international_reputation      work_rate  pace  \
1               3            3                         5    High/Medium  89.0   
2               5            4                         5    High/Medium  72.0   
9               3            1                         5  Medium/Medium   NaN   
10              4            3                         4    Medium/High  58.0   
11              4            2                         4    Medium/High  62.0   
...           ...          ...                       ...            ...   ...   
179683          3            2                         1  Medium/Medium  65.0   
179693          3            1                         1  Medium/Medium   NaN   
179720          3            2                         1  Medium/Medium  66.0   
179783          2            2                         1  Medium/Medium  65.0   
179819          3            2                         1    Medium/High  58.0   

        shooting  passing  dribbling  defending  physic  attacking_crossing  \
1           93.0     66.0       80.0       45.0    88.0                  47   
2           88.0     94.0       87.0       65.0    78.0                  95   
9            NaN      NaN        NaN        NaN     NaN                  17   
10          73.0     80.0       80.0       85.0    84.0                  63   
11          39.0     66.0       69.0       89.0    87.0                  54   
...          ...      ...        ...        ...     ...                 ...   
179683      50.0     41.0       53.0       25.0    32.0                  52   
179693       NaN      NaN        NaN        NaN     NaN                  25   
179720      51.0     34.0       52.0       29.0    46.0                  30   
179783      41.0     55.0       54.0       37.0    42.0                  44   
179819      40.0     46.0       42.0       47.0    60.0                  38   

        attacking_finishing  attacking_heading_accuracy  \
1                        96                          83   
2                        85                          55   
9                        13                          29   


Looking at the data now there are 2 columns that would be useful but need to be altered for a machine learning model:
* player_positions - change to main position with options: forward (FW), midfielder (MD), defender (DF), goalkeeper (GK)

* work_rate - split into 2 columns repensenting attacking work-rate & defensive work-rate respectively.

In [54]:
#----------------------------------
# player_positions -> main_position
#----------------------------------

def calc_main_position(positions: str):
    # list of all outfield positons can be found at https://www.fifplay.com/encyclopedia/position/
    forward_positions = [
        'ST', 
        'CF',
        'RF',
        'LF',
        'RW',
        'LW',
    ]

    midfielder_positions = [
        'CM',
        'CDM',
        'CAM',
        'RM',
        'LM',
    ]

    defender_positions = [
        'CB',
        'RB',
        'LB',
        'RWB',
        'LWB',
    ]

    primary_position = positions.split(',')[0] # using first mentioned position as primary position

    if primary_position in forward_positions:
        return 'FW'
    elif primary_position in midfielder_positions:
        return 'MD'
    elif primary_position in defender_positions:
        return 'DF'
    elif primary_position == 'GK':
        return 'GK'
    else:
        return None # no valid position found

# create new column with function applied to each value
df['main_position'] = df['player_positions'].apply(lambda pos: calc_main_position(pos))
# show new column
print(df['main_position'])
# check for nulls
print('\nNulls: ', df['main_position'].isnull().sum())
# drop old column
df.drop(['player_positions'] ,axis=1,inplace=True)



1         FW
2         MD
9         GK
10        MD
11        DF
          ..
179683    FW
179693    GK
179720    FW
179783    MD
179819    MD
Name: main_position, Length: 9371, dtype: object

Nulls:  0


C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\2651282946.py:43: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['main_position'] = df['player_positions'].apply(lambda pos: calc_main_position(pos))
C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\2651282946.py:49: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(['player_positions'] ,axis=1,inplace=True)


In [64]:
#-------------------------------------------------------
# work_rate -> attacking_work_rate & defending_work_rate
#-------------------------------------------------------

# work_rate format: '[attacking work rate] / [defending_work_rate]'
def calc_att_work_rate(work_rate: str):
    return work_rate.split('/')[0]

def calc_def_work_rate(work_rate: str):
    return work_rate.split('/')[1]

# create new attacking and defending work rate columns
df['attacking_work_rate'] = df['work_rate'].apply(lambda work_rate: calc_att_work_rate(work_rate))
df['defending_work_rate'] = df['work_rate'].apply(lambda work_rate: calc_def_work_rate(work_rate))
# drop old work_rate column
df.drop(['work_rate'], axis=1, inplace=True)

C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\1624192788.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['attacking_work_rate'] = df['work_rate'].apply(lambda work_rate: calc_att_work_rate(work_rate))
C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\1624192788.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['defending_work_rate'] = df['work_rate'].apply(lambda work_rate: calc_def_work_rate(work_rate))
C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\1624192788.py:16: SettingWithC

Now I will sort each column into categorical & numerical statistical types.

Categorical:
* player_id
* nationality_id
* club_team_id
* league_id
* league_level
* club_jersey_number
* club_cont


Now I reduce the amount of data we have by dropping the columns that are not relevant (not performance or value related).

As I are looking to predict the value of a footballer, I need to find a column which quantifies an individual's value to be the target.

In this case value_eur or wage_eur accurately fit this requirement.

I will check if these columns have any null values.

In [56]:
wage_eur_n_nulls = df['wage_eur'].isnull().sum()
wage_eur_n_nulls

0

In [57]:
value_eur_n_nulls = df['value_eur'].isnull().sum()
value_eur_n_nulls

12

In [58]:
n_rows = df.shape[0] # amount of rows
n_rows

9371

In [59]:
value_eur_n_nulls / n_rows * 100 # percentage nulls

0.12805463664496852

wage_eur has 0 null values but value_eur has 12. 

This is not a lot of null values at only 0.12% but I still need to deal with these nulls.

To deal with these null values I will omit them from the data by dropping the rows. 

In [60]:
df.dropna(subset=['value_eur'], inplace=True)
df.shape[0]

C:\Users\nalei76\AppData\Local\Temp\ipykernel_22036\662795726.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.dropna(subset=['value_eur'], inplace=True)


9359

# Exploring the data relationship